### PR TITLE
Make `visit_clobber` panic-safe.

### PR DESCRIPTION
**Pull request by nnethercote**
Local measurements indicate the performance effect is negligible.

r? @petrochenkov

### CODE REVIEW

**Comment by nnethercote** [comment]
Sorry, I meant to say this in the description:

r? @petrochenkov 

(I have updated the description to match that.)

**Comment by petrochenkov** [comment]
Will this properly report ICEs if they happen inside `f`?
Could you show what the message is if the panic actually happens.

File: src/libsyntax/mut_visit.rs [hunk 1]
```diff
@@ -309,7 +310,12 @@ pub trait MutVisitor: Sized {
 //
 // No `noop_` prefix because there isn't a corresponding method in `MutVisitor`.
 pub fn visit_clobber<T, F>(t: &mut T, f: F) where F: FnOnce(T) -> T {
-    unsafe { std::ptr::write(t, f(std::ptr::read(t))); }
+    unsafe {
```
Comment (Centril) [inline src/libsyntax/mut_visit.rs]: Could you leave a note as to why it is safe?
(we should really make tidy require this at some point... it's troubling that this information is lacking in the compiler of all places...)

**Comment by nnethercote** [comment]
> Could you show what the message is if the panic actually happens.

@petrochenkov: It shows the panic message like you'd expect. See [here](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=557a4f15b512fa41f3b265258d6ac98a) for an example.

File: src/libsyntax/mut_visit.rs (hunk 1 above)
Comment (nnethercote) [reply src/libsyntax/mut_visit.rs]: I added a note. I wasn't sure how to explain it, suggestions are welcome.

**Pull request merged**